### ISSUE TITLE
For what are the use of the POST urls curl -X POST https://backstroke.us/_dfasdf1asd1f3dsf32df...?

### ISSUE THREAD

**Issue by evandrocoan** [1egoman/backstroke#62]
Accordingly to:

1. https://github.com/1egoman/backstroke/issues/59 How to configure webhooks on my forks?

I created a python script to loop through all my forks call the url 
`curl -X POST https://backstroke.us/_dfasdf1asd1f3dsf32df...`. 

Thought, they are on my git modules, as follows. Can they be public or do I need to remove them?
```
[submodule "Packages/ANSIescape"]
	path = Packages/ANSIescape
	url = https://github.com/evandrocoan/SublimeANSI
	upstream = https://github.com/aziz/SublimeANSI
	backstroke = https://backstroke.us/_597befd29102c500...
```
1. https://github.com/evandrocoan/SublimeTextStudio/blob/develop/.gitmodules#L126-L130

This is the script check part:
```python
upstream   = configParser.get( section, "upstream" )
backstroke = configParser.get( section, "backstroke" )
path       = configParser.get( section, "path" )

# log( 1, upstream )
# log( 1, backstroke )

# https://stackoverflow.com/questions/2018026/what-are-the-differences-between-the-urllib-urllib2-and-requests-module
if len( backstroke ) > 20:

    # https://stackoverflow.com/questions/28396036/python-3-4-urllib-request-error-http-403
    req = urllib.Request( backstroke, headers={'User-Agent': 'Mozilla/5.0'} )
    res = urllib.urlopen( req )

    # https://stackoverflow.com/questions/2667509/curl-alternative-in-python
    print( res.read() )
```
1. https://github.com/evandrocoan/SublimeTextStudioChannel/blob/master/update.py

**Comment by 1egoman** [comment]
@evandrocoan That url is the url that is called behind the scenes by Backstroke (ie, the github webhook). As explained on the dashboard, running `curl -X POST https://backstroke.us/_XXX...` kicks off Backstroke at any arbitrary time if you'd like to run the check external to Backstroke for some reason.

**Comment by evandrocoan** [comment]
But if I understand correctly, backstroke is kicking off automatically. Why would I need the URL?

If some one take my URL on the git repository they can start kicking it like ddos, is there a problem? Should the URL be a secret or they can be public?

**Comment by 1egoman** [comment]
@evandrocoan In 99% of cases, the url is not required. I've run into cases (and others have as well) where they've wanted to integrate backstroke into a different system and wanted to kick off Backstroke on their own. That's the only reason why the url is user facing (it's hidden in the dashboard to try and emphasize this too).

As fas as any denial of service concerns: the url can be public. Since nothing will happen if the repo is already up-to-date, the worst thing that could happen is that your repository is checked more often for changes. Also, backstroke.us is behind cloudflare, which should take care of any abuse.

**Comment by evandrocoan** [comment]
> the url can be public

I got this response from GitHub while running my pull algorithm:

`API rate limit exceeded for evandrocoan.` 

How does it does my name? Is it from the consecutive calls to the backstroke URL `https://backstroke.us/_dfasdf1asd1f3dsf32df...`?

If so, could someone be using my URL to kick off GitHub, then when I actually need to kick some repository, I cannot because they exhausted/kick off all my daily quota?

```
...
[DEBUG] 13:14:24:617565  -5069 Index: 28/54, submodule "Packages/Indent and braces"
b'{"status":"ok","output":{"status":"ok","pullRequest":{"msg":"There\'s already a pull request for this repo, no need to create another."},"isEnabled":true,"many":false,"forkCount":1}}'
[DEBUG] 13:14:25:753405 135840 Index: 29/55, submodule "Packages/Invert Selection"
b'{"status":"ok","output":{"status":"ok","pullRequest":{"msg":"There\'s already a pull request for this repo, no need to create another."},"isEnabled":true,"many":false,"forkCount":1}}'
[DEBUG] 13:14:26:907257 153852 Index: 30/56, submodule "Packages/LaTeXTools"

ERROR!  b'{"error":{"code":403,"status":"Forbidden","message":"{\\"message\\":\\"API rate limit exceeded for evandrocoan.\\",\\"documentation_url\\":\\"https://developer.github.com/v3/#rate-limiting\\"}"}}'
```

**Comment by 1egoman** [comment]
@evandrocoan Backstroke uses your Github access token to make requests, and Github rate limits users with tokens to 6000 requests per hour. At worst, someone would prevent Backstroke from working on your repositories for an hour, and then once the rate limit expired, Backstroke would work again. This is a Github limitation and something that cannot be easily circumvented.

**Comment by evandrocoan** [comment]
On the example above, I did not hit the 6000 request limit, I only had make about 50 sequential requests to the URL. GitHub seems to block repeated requests, on the 6000 hit limit.

So, by the fact that someone has my URLs they can be popping the GitHub API under my name, therefore run dry all my 6000 requests?

**Comment by 1egoman** [comment]
Backstroke makes many calls to Github for a given request. If someone were to hit a link's webhook url repeatedly over an over, they would "run dry" all of your requests. At worst, this means that your repository would be out of date for an hour until the Github rate limit reset.

**Comment by evandrocoan** [comment]
Thanks for the patience. Seeing now how it works, I would like to ask if it is possible to block repeated requests to the same URL, as the GitHub does their rate limit.

So as my URLs are public, anyone can create a simple script which start pinging repeatedly the same URL, over and over again for ever. This would cause GitHub to exhaust all my 6000 requests just for on repository URL, and can keep me blocking for ever from ping other repositories, as they are kicking off all my requests limit with the same repository.

Basically limiting the kicking off of a URL as only one request is acceptable for 6000 seconds. So it would block any one from exhausting my GitHub rate limit using one or just a few URLs from my account.

**Comment by 1egoman** [comment]
Currently, this doesn't exist. If you'd like to try to implement it, [here's the most up-to-date branch](https://github.com/1egoman/backstroke/tree/fix-it-all). I can review any submitted pull requests.

**Comment by evandrocoan** [comment]
Thanks, for now I cannot engage on it. So the simplest solution seems to be removing all the URL from public. 

But now I we can discuss the algorithm to accomplish so, I can think about one, do you approve it for implementation?

1. Implement a hash table with the repositories as the key, so when performing check to the URL `https://backstroke.us/_dfasdf1asd1f3dsf32df`, this table `dfasdf1asd1f3dsf32df` is checked and the time stamp is accessed and compared against the current time in EPOCH as 1501635118. 

    If there is the minimum time of 6000 seconds passed since the last input, the call is approved and the time updated on the hashtable. Otherwise the call is just rejected. But when there is not entry, a entry is added and the call is performed.

    This algorithm can be improved, so the hashtable is restarted every 24h, to clean/free up some memory, and also when its size passes some limit to avoid staving the server memory.
